caffe2/utils/threadpool/pthreadpool-cpp.cc

```diff
@@ -0,0 +1,75 @@
+#include <caffe2/utils/threadpool/pthreadpool-cpp.h>
+#include <c10/util/Exception.h>
+
```

Comment:
All of the pthreadpool symbols present in this file, namely pthreadpool_create, pthreadpool_destroy, pthreadpool_get_threads_count, and pthreadpool_parallelize_1d have a C2 and third party implementation.  The build system picks which one to use.  This implementation is oblivious to that.

Reply:
Thanks for this comments Ashkan. They make it a little easier to understand. I shall adopt this :).

Reply:
> > That seems to have given C2's implementation the upper hand in performance, hence justifying the added maintenance complexity, until the third party version improved in parallel surpassing the efficiency of C2's implementation as I have verified in benchmarks.
> 
> Regarding this, do we understand what gives better performance for OSS pthreadpool? Is it the use of futex vs std::conditional_variable (not sure if that is implemented using futex)?

I think it is just a bug in C2's implementation but I am not certain.  That's just my guess.